governance/program/tests/process_cast_vote.rs

```diff
@@ -0,0 +1,522 @@
+}
+
+#[tokio::test]
+async fn test_cast_vote_with_threshold_below_50_and_vote_not_tipped() {
+    // Arrange
+    let mut governance_test = GovernanceProgramTest::start_new().await;
+
+    let realm_cookie = governance_test.with_realm().await;
+    let governed_account_cookie = governance_test.with_governed_account().await;
+
+    let mut governance_config =
+        governance_test.get_default_governance_config(&realm_cookie, &governed_account_cookie);
+
+    governance_config.vote_threshold_percentage = 40;
+
+    let mut account_governance_cookie = governance_test
+        .with_account_governance_using_config(
+            &realm_cookie,
+            &governed_account_cookie,
+            &governance_config,
+        )
+        .await
+        .unwrap();
+
+    let token_owner_record_cookie = governance_test
+        .with_initial_community_token_deposit(&realm_cookie)
+        .await;
+
+    // Total 210 tokens
+    governance_test
+        .mint_community_tokens(&realm_cookie, 110)
+        .await;
+
+    let proposal_cookie = governance_test
+        .with_signed_off_proposal(&token_owner_record_cookie, &mut account_governance_cookie)
+        .await
+        .unwrap();
+
+    // Act
+    governance_test
+        .with_cast_vote(&proposal_cookie, &token_owner_record_cookie, Vote::Yes)
+        .await
+        .unwrap();
+
+    // Assert
+
+    let proposal_account = governance_test
+        .get_proposal_account(&proposal_cookie.address)
+        .await;
+
+    assert_eq!(ProposalState::Voting, proposal_account.state);
+}
+
+#[tokio::test]
+async fn test_cast_vote_with_voting_time_expired_error() {
+    // Arrange
+    let mut governance_test = GovernanceProgramTest::start_new().await;
+
+    let realm_cookie = governance_test.with_realm().await;
+    let governed_account_cookie = governance_test.with_governed_account().await;
+
+    let mut account_governance_cookie = governance_test
+        .with_account_governance(&realm_cookie, &governed_account_cookie)
+        .await
+        .unwrap();
+
+    let token_owner_record_cookie = governance_test
+        .with_initial_community_token_deposit(&realm_cookie)
+        .await;
+
+    let proposal_cookie = governance_test
+        .with_signed_off_proposal(&token_owner_record_cookie, &mut account_governance_cookie)
+        .await
+        .unwrap();
+
+    let proposal_account = governance_test
+        .get_proposal_account(&proposal_cookie.address)
+        .await;
+
+    let vote_expired_at_slot = account_governance_cookie.account.config.max_voting_time
+        + proposal_account.voting_at.unwrap()
+        + 1;
+    governance_test
+        .context
+        .warp_to_slot(vote_expired_at_slot)
+        .unwrap();
+
+    // Act
+
+    let err = governance_test
+        .with_cast_vote(&proposal_cookie, &token_owner_record_cookie, Vote::No)
+        .await
+        .err()
+        .unwrap();
+
+    // Assert
+
+    assert_eq!(err, GovernanceError::ProposalVotingTimeExpired.into());
+}
```

Comment:
Can you add a test trying to vote twice?

Reply:
Added the test and also more specific error VoteAlreadyExists if that happens